MDEV-35254 Make iterations count configurable in PARSEC plugin

deo002 opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-35254*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description
Currently it is not possible with the PARSEC plugin to define the number of iterations to be used when generating the key corresponding to the password.

To ensure proper security for users long term, the PARSEC plugin should allow specifying the number of iterations. Additionally, the default should be something more secure.

This patch adds a global plugin variable parsec_iterations to define define the number of iterations to be used when generating the key corresponding to the password. It has a default value, lower and upper bounds.

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
-->
## Basing the PR against the correct MariaDB version
- [X] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and coding style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [X] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [X] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

deo002 committed:
MDEV-35254 Make iterations count configurable in PARSEC plugin

This patch adds a global and session level plugin variable
parsec_iterations to define the number of iterations to be
used when generating the key corresponding to the password.
It has a default value, lower and upper bounds.

deo002 commented:
Related PR: https://github.com/mariadb-corporation/mariadb-connector-c/pull/297

gkodinov reviewed (dismissed):
Please make sure the tests are passing. There are some parsec related failures as a result of your changes in buildbot.

gkodinov reviewed (commented):
OK, I'll leave that to the final reviewer. Still learning how CI works. OK as far as the preliminary review goes.

grooverdan commented on mysql-test/suite/plugins/t/parsec_iterations.test:
```diff
@@ -0,0 +1,86 @@
+#
+# Test for MDEV-35254: PARSEC plugin should allow DBAs to specify number of iterations
```
note this could have been included in the existing test case `parsec.test`.

Tests would normally echo this line rather than it being a comment.

grooverdan commented on mysql-test/suite/plugins/t/parsec_iterations.test:
```diff
@@ -0,0 +1,86 @@
+drop user t_iter@'%';
+
+-- echo #
+-- echo # Preserve the state that existed before the test case was executed
```
no need to echo non-test information.

grooverdan commented on mysql-test/suite/plugins/t/parsec_iterations.test:
```diff
@@ -0,0 +1,86 @@
+-- echo # Preserve the state that existed before the test case was executed
+-- echo #
+
+set global parsec_iterations = 262144;
```
we'd normally save the default global value as  user variable at the beginning of the test and set it back at the end. This enables a change in default without changing the test case.

grooverdan replied:
An end of tests would normally include a `--echo # End of 12.4 tests` statement to make the test case easier to merge if in future tests are added in later versions.

kulebelden-wq commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +38,52 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
+
+
+static void update_parsec_iterations(MYSQL_THD thd,
+              struct st_mysql_sys_var *var  __attribute__((unused)),
```
even me that's what i thought in the first place

FooBarrior replied:
So we need this gcc-ism

vuvova replied:
do we? I thought it's standard C++ and a common pattern we already use. What build issues?

FooBarrior commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +38,52 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
+static MYSQL_SYSVAR_UINT(iterations, iterations, PLUGIN_VAR_NOCMDOPT,
+       "Number of iterations to be used when generating the key corresponding to the password",
+       NULL, update_parsec_iterations, PARSEC_ITERATIONS_DEFAULT, PARSEC_ITERATIONS_MIN,
+       PARSEC_ITERATIONS_MAX, PARSEC_ITERATIONS_STEP);
```
Can PARSEC_ITERATIONS_STEP be anything but 1? If not, then the existence of this variable confuses more, than if it was just "1" here

FooBarrior commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +38,52 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
+      ME_WARNING, iterations);
+}
+
+static MYSQL_SYSVAR_UINT(iterations, iterations, PLUGIN_VAR_NOCMDOPT,
```
Why NOCMDOPT? I think it can be perfectly set with a command line.

FooBarrior replied:
Yes, that's pretty surprising for me as well. Let me check..

FooBarrior commented on mysql-test/suite/plugins/r/parsec.result:
```diff
@@ -32,7 +32,7 @@ drop user test1@'%';
 show grants for test2@'%';
 Grants for test2@%
-GRANT USAGE ON *.* TO `test2`@`%` IDENTIFIED VIA parsec USING 'P0:salt:password'
+GRANT USAGE ON *.* TO `test2`@`%` IDENTIFIED VIA parsec USING 'P8:salt:password'
```
I doubt this should have been changed.

FooBarrior commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +38,52 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
 
+constexpr unsigned int PARSEC_ITERATIONS_DEFAULT= 1u << 18;
+constexpr unsigned int PARSEC_ITERATIONS_MIN= 1u << 10;
+constexpr unsigned int PARSEC_ITERATIONS_MAX= 1u << 19;
```
this isn't a correct max value. To stay on the safe side, make it 1<<31, and that'll be our technical limitation of 32-bit int (because of my_bit.h function)

FooBarrior replied:
> After '9', we have some symbols like ':', ';'

exactly.

FooBarrior replied:
Oh, my bad! Our format
XY:salt:hash

Allows storing an arbitrary number in y. That is, for X=P, that is PBKDF2, and unadjusted iterations power Y=27, it'll be P27:salt:hash

deo002 replied:
> Oh, my bad! Our format XY:salt:hash
> 
> Allows storing an arbitrary number in y. That is, for X=P, that is PBKDF2, and unadjusted iterations power Y=27, it'll be P27:salt:hash

I think 27 might be a slip - shouldn't it be 21(31 - 10)?

FooBarrior commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -204,10 +250,11 @@ int digest_to_binary(const char *hash, size_t hash_length,
   if (hash_length != sizeof (*stored) || *out_length < sizeof(*memory) ||
       stored->algorithm != 'P' ||
-      stored->iterations < '0' || stored->iterations > '3' ||
+      stored->iterations < '0' || stored->iterations > ITER_MAX_VAL ||
```
This is a question to me rather. And I don't remember the reason:) Perhaps a dumb sanity check.

FooBarrior reviewed (commented):
Good start!
Lots of comments, enough for now, then we'll do another iteration

FooBarrior reviewed (changes requested):
Forgot to press "request changes"

deo002 commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +38,52 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
 constexpr size_t CLIENT_RESPONSE_LENGTH= CHALLENGE_SCRAMBLE_LENGTH
                                          + ED25519_SIG_LENGTH;
 
+constexpr unsigned int PARSEC_ITERATIONS_DEFAULT= 1u << 18;
```
Yes, it was 1024. I thought of keeping it close to the OWASP's recommended value for PBKDF2-HMAC-SHA512(210,000) as stated in the Jira ticket.

FooBarrior reviewed (commented):
Forgot to press "request changes"

gkodinov reviewed (approved):
For the sake of process, I'd like to mark the preliminary review as done. I see that the final review is in full swing. Please keep working with the other reviewers.

vuvova commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +40,90 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
+/*
+  copied from my_bit.h and m_string.h because making a service
+  will be an overkill
+*/
```
why not include `my_bit.h`? it looks like a clean header, no dependencies

deo002 replied:
ABI may break if the my_bit.h's definitions change between versions. Although, it is extremely unlikely that they will ever change. Will add it as a header.

deo002 replied:
My bad, ABI is not a concern here. All are inline functions, compiled into the caller.

deo002 replied:
Tried including my_bit.h directly (requires my_global.h as a prerequisite). This caused a Windows build failure. Any suggestions on how should I proceed? @vuvova

vuvova replied:
no, don't include `my_global.h`, if you cannot use `my_bit.h` without it then better revert to local functions.

But why `my_bit.h` needs `my_global.h` ? if it's only because of uchar/uint/ulonglong then you can either
* typedef them in the plugin, or
* change `my_bit.h` to use `unsigned char`, `unigned int`, etc., making it independent from `my_global.h`

deo002 replied:
Got it! Thanks!

deo002 replied:
I made my_bit.h standalone. I changed to unsigned char, unsigned int etc, but also had to change some macros like C_MODE_START, C_MODE_END, CONSTEXPR. Please let me know if this change is ok.

FooBarrior replied:
Parsec already uses ma_global.h from libmariadb. Perhaps it's fine to use it here either?
my_global.h lies in the same directory as my_bit.h: in root/include. If my_bit.h is fine to include, then why my_global.h is not?

vuvova commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +40,90 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
+  parsec_iterations_t iterations_user_input= *static_cast<const parsec_iterations_t *>(save);
+  parsec_iterations_t iterations= parsec_round_up_to_next_power(iterations_user_input);
+  if (iterations != iterations_user_input)
+    my_printf_error(ER_WRONG_VALUE_FOR_VAR, "parsec: parsec_iterations rounded up to %d (next supported value)",
```
remove "parsec: " prefix, it's redundant. And may be " (next supported value)" too, because the variable help text should say that, not a warning message after the fact

deo002 commented:
Working on it!

FooBarrior commented on include/my_bit.h:
```diff
@@ -17,13 +17,29 @@
 */
 
-C_MODE_START
+MY_BIT_C_MODE_START
```
MY_BIT_C_MODE_START is weird with this MY_BIT prefix. There is nothing specific to MY_BIT in this macro

If you want to avoid duplicate macro definition warning, then either:
1. `#undef` macro before `#define` -- undeffing an undefined macro doesn't produce a new warning
2. Wrap the definition in `#ifndef/#endif`

The first solution is one line shorter, but if the definition would change in my_global someday, it would likely not be updated here, and those who are using my_bit.h would end up in stale definition (hiding the one from my_global.h).

So I'd go with `#1`, however, see my next comment.

vuvova replied:
I'd do
```c
#ifdef __cplusplus
extern "C" {
#else
#define constexpr
#endif

// all my_bit.h content

#ifdef __cplusplus
}
#else
#undef constexpr
#endif
```

FooBarrior replied:
And also uint, uchar,...

And do you suggest to change CONSTEXPR to constexpr?

vuvova replied:
no uint, no uchar, this is already replaced by `unsigned int`, `unsigned char`, etc. only c/c++ changes to make a universal c/c++ header.

FooBarrior commented on include/my_bit.h:
```diff
@@ -91,17 +107,17 @@ Count bits in 32bit integer
  (Original code public domain).
 */
-static inline uint my_count_bits_uint32(uint32 v)
+static inline unsigned int my_count_bits_uint32(unsigned int v)
```
@vuvova that's wrong: my_bit is still *my_*. that is, from my_ namespace and ecosystem. Every other component from my_ ecosystem still uses my_bit, and with this chang they will end up with using semantically incompatible values. Yes, c++ allows assigning unsigned long long to ulonglong, but that's semantically incorrect, and when someone does that, they should understand it's a hack, and should understand why they do that (for example , it could be a typecast from one ecosystem to another).

I'm fine with both using my_global.h, I saw it's being used in other plugins, f.ex. auth_pam.c (they even have it twice :D)

Maybe the naming is bad, and I myself intuitively avoid using anything that is global (perhaps, that is why innodb has "universal" rather), but the contents of my_global.h look fine to me, maybe a bit more macros that we need, but nothing SQL-specific or suspicuous.

I am also fine to keep a small set of functions copied and stored independently in a plugin. I like the idea of keeping them standalone.

vuvova replied:
ulonglong is a shortcut alias for `unsigned long long`, they're completely interchangeable. `my_global.h` is not for plugins, it's contains too much stuff that only server needs. `my_bit.h` is a small header-only library with no external dependencies, let's keep it that way

FooBarrior replied:
Ok, I'm fine with having my_bit.h a standalone library/header, the naming is perhaps confusing then. But ok.

> ulonglong is a shortcut alias for unsigned long long

Now this is a separate, sleeky point... I couldn't easily prove or otherwise rebut it.
I was hoping to find something in CODING_STANDARDS.md to cite, but it barely mentions these types. 
The only particle relevant to this discussion is
> The usage of types long and unsigned long (**and its ulong alias**) 

Though it doesn't clarify the issue. Calling it an alias doesn't mean it's supposed to be used interchangeably: my_off_t is also a type alias.  

Note that the function declarations would be actually incompatible:
```
int f (ulonglong);
int (*g)(unsigned long long) = f;
```
Will produce an error because `ulonglong` is `unsigned long`. That's fixable, one should just rewrite `ulonglong` definition as `unsigned long long`.

FooBarrior commented on include/my_bit.h:
```diff
@@ -17,13 +17,29 @@
 #ifndef MY_BIT_INCLUDED
 #define MY_BIT_INCLUDED
 
+/* C vs C++ */
```
It would have to be a separate commit

FooBarrior commented on include/my_bit.h:
```diff
@@ -91,17 +107,17 @@ Count bits in 32bit integer
 
 
-static inline uint my_count_bits(ulonglong x)
+static inline unsigned int my_count_bits(unsigned long long int x)
```
@vuvova one more option: extract the bare minimum required to use such headers like my_bit.h into a new header: my_... types? But we'll also need that C_MODE_START and whatever. my_minimum.h?

Include that header both into my_global.h and my_bit.h, so that people wouldn't have to make two inclusions when using them, nor manage the order of inclusions (now my_global should always go first)

Put both in mysys. Currently. we have mysys libs using my_bit.h, and my_priv.h includes my_global.h directly, which cycles the component dependencies.

FooBarrior replied:
> Put both in mysys. Currently. we have mysys libs using my_bit.h, and my_priv.h includes my_global.h directly, which cycles the component dependencies.

Oh, I realized that all mysys headers are actually in root/include :) So I'd effectively count my_global.h and my_bit.h as part of mysys, just stored differently.

vuvova replied:
no, `my_bit.h`, again, is a small clean header-only library with no dependencies. don't contaminate with with `my_global.h` or mysys or anything. it's completely independent and doesn't need anything.

FooBarrior reviewed (changes requested):
Expressed my concerns, basically directed to @vuvova.

deo002 committed:
MDEV-35254 refactor: Make my_bit.h an independent header

Removed type aliases and macros defined in my_global.h from my_bit.h,
making it possible to include it in plugins or external code without
pulling in my_global.h

deo002 commented:
I have pushed the requested changes. Some tests are failing, they don't seem to be related to this PR.

FooBarrior commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -27,6 +27,7 @@
 #include <mysqld_error.h>
 #include <mysql/service_thd.h>
 #include "scope.h"
+#include "my_bit.h"
```
this is a wrong commit for parsec changes.

Put the refactoring commit first in your branch, then parsec feature on top.

Note that we don't have a requirement to follow _conventional commit_ notation.

FooBarrior commented:
@deo002 Also put MDEV-35254 in the refactor commit header, that'll make it easier for us to cherry-pick changes if we'll want to.

deo002 commented:
Done!

deo002 committed:
MDEV-35254 Add get_current_thd() service function

vuvova commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +40,42 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
 constexpr size_t CLIENT_RESPONSE_LENGTH= CHALLENGE_SCRAMBLE_LENGTH
                                          + ED25519_SIG_LENGTH;
 
+constexpr parsec_iterations_t ITER_FACTOR_BASE_VAL= 10u;
+constexpr parsec_iterations_t ITER_BASE_VAL= 1u << ITER_FACTOR_BASE_VAL;
+constexpr parsec_iterations_t PARSEC_ITERATIONS_MAX= 1u << 31;
+
```
looks valid

vuvova commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -38,6 +40,42 @@ constexpr size_t PBKDF2_HASH_LENGTH= ED25519_KEY_LENGTH;
+const char parsec_dig_vec_base62[] =
+  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+static void update_parsec_iterations(MYSQL_THD thd,
+              struct st_mysql_sys_var *var, void *var_ptr, const void *save)
+{
+  parsec_iterations_t iterations_user_input= *static_cast<const parsec_iterations_t *>(save);
+  parsec_iterations_t iterations= my_round_up_to_next_power(iterations_user_input);
+  if (iterations != iterations_user_input)
+    my_printf_error(ER_WRONG_VALUE_FOR_VAR, "parsec_iterations rounded up to %d",
+      ME_WARNING, iterations);
+  *static_cast<parsec_iterations_t*>(var_ptr)= iterations;
+}
```
nitpicking. signed/unsigned point is moot if we use a lower MAX value per previous comment

vuvova commented on plugin/auth_parsec/server_parsec.cc:
```diff
@@ -198,16 +237,19 @@ int hash_password(const char *password, size_t password_length,
       stored->algorithm != 'P' ||
-      stored->iterations < '0' || stored->iterations > '3' ||
+      stored->iterations < '0' || stored->iterations > ITER_MAX_VAL ||
       stored->colon != ':' || stored->colon2 != ':')
```
looks valid too

vuvova committed:
MDEV-35254 post-review fixes

* reduced PARSEC_ITERATIONS_MAX because PKCS5_PBKDF2_HMAC() takes
  a signed int
* instead of checking for '0'..'L' range (which includes punctuation,
  check for the valid result of base62_to_uchar()
* add a connection test with non-default parsec_iterations
* make parsec_iterations a normal command line option
* update C/C to allow larger number of iterations in the client parsec

vuvova reviewed (approved):
Ok, this looks good.

The service should be in a separate commit. Copilot had a couple of good comments.
And an test with an actual connection attempt was missing.

I've rebased on top of main and fixed the above, now it's good for 13.1 and testing.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=4504) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=4504) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: deo002<br/>:x: vuvova<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=4504) it.</sub>

deo002 commented:
On it! Will push the requested changes soon!
P.S. Just checked, it's all done! Thanks for the fixes and refactor @vuvova

vuvova commented:
Approved by testing (see [MDEV-35254](https://jira.mariadb.org/browse/MDEV-35254)) I'm merging it now
